fix(eslint-plugin): suggest object, not Record<string, never>

kmin-jeong opened the pull request:
<!--
👋 Hi, thanks for sending a PR to typescript-eslint! 💖
Please fill out all fields below and make sure each item is true and [x] checked.
Otherwise we may not be able to review your PR.
-->

## PR Checklist

- [ ] Addresses an existing open issue: fixes #5949
- [ ] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [ ] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

<!-- Description of what is changed and how the code change does that. -->

Josh-Cena commented:
Are we un-suggesting `Record<string, never>` or `Record<string, unknown>`? To me the former is more problematic, but this PR is not changing the right thing.

kmin-jeong commented:
@Josh-Cena @JoshuaKGoldberg  I changed the part of the message according to the proposal in the issue and sent a pr draft. I saw the issue and understood that the object was used instead of 'Record<string, unknown>'. Did I get it wrong? I haven't worked on the code and the test code yet.

bradzacher commented:
This is the correct fix (it's just a re-creation of #5018).

kmin-jeong commented:
@bradzacher @Josh-Cena @JoshuaKGoldberg 
I fixed the part in the record. I add the short test code, and converted the draft into a pr. If I understand something wrong or there is something wrong with the code, I will correct it.

bradzacher reviewed (changes requested):
looks like you've introduced some type errors

kmin-jeong commented:
@bradzacher I add type Types and type checkrequiredTypes to fix lint errors.

bradzacher commented on packages/eslint-plugin/src/rules/ban-types.ts:
```diff
@@ -3,16 +3,7 @@ import { AST_NODE_TYPES } from '@typescript-eslint/utils';
 
 import * as util from '../util';
 
-type Types = Record<
-  string,
-  | null
-  | false
-  | string
-  | {
-      message: string;
-      fixWith?: string;
-    }
->;
```
I'm not sure I understand - how come you changed this type here? It doesn't seem like this was required for your PR?

kmin-jeong replied:
I understood that changing record <string, never> to object. So I changed it to types and used it.

kmin-jeong replied:
I modified The type types back to record<>.

kmin-jeong commented:
After looking at this issue and the discussion mentioned in the issue, I added the test code and requested the pr. 
I don't know the range where I shouldn't use type Record <string, never>. if I just need to modify the message  to 'use object instead', or if I need to modify the code in the  type Record <string, never> and let the message 'use object instead' come out. What do you think of @JoshuaKGoldberg who posted this issue and @bradzacher who posted a review? If you think it's unnecessary pr, I'll close it.

bradzacher commented:
@kmin-jeong the fix we're looking for here is to update the message so that it suggests using `object`.
You're first commit (f679c9950b36ed0e947d3855986bfb899b72a44e) was the correct fix!

kmin-jeong commented:
@bradzacher I vaguely thought I should fix the code, too. I modified ban-types.ts again like Commit (https://github.com/typescript-eslint/typescript-eslint/commit/f679c9950b36ed0e947d3855986bfb899b72a44e).

JoshuaKGoldberg commented:
Psst - you don't have to keep @ tagging us. We'll see the updates come in and reply when we can. 🙂

JoshuaKGoldberg commented on packages/eslint-plugin/tests/rules/ban-types.test.ts:
```diff
@@ -388,6 +388,39 @@ let b: Foo<NS.Good>;
         },
       ],
     },
+
+    {
+      code: noFormat`
+declare let j: unknown;
+let m3 = { ...(j as Record<string, never>) };
+      `,
+      output: `
+declare let j: unknown;
+let m3 = { ...(j as object) };
+      `,
+      options: [
+        {
+          types: {
+            'Record<string, never>': {
+              message: 'Use object instead.',
+              fixWith: 'object',
+            },
+          },
+        },
+      ],
+      errors: [
+        {
+          messageId: 'bannedTypeMessage',
+          data: {
+            name: 'Record<string,never>',
+            customMessage: ' Use object instead.',
+          },
+          line: 3,
+          column: 21,
+        },
+      ],
+    },
+
```
There's no need for this test. It isn't testing anything new - this file already tests that providing `types` options are respected.

```suggestion

```

JoshuaKGoldberg reviewed (approved):
The code changes look good, thanks! I'll take care of the test revert.

I also filled out the `Overview` in the PR.

Josh-Cena commented:
Shouldn't the PR title say "instead of `Record<string, unknown>` then?

JoshuaKGoldberg commented:
Ah thanks, good catch on the title - updated.

And yeah they're not _quite_ the same but `object` is close enough for most cases.
